Michigan StateCleaner

necabotheking opened the pull request:
Creates the draft pull request for the Michigan StateCleaner Class

Issues:
- generating UUIDs for full_names of organizations and individuals is not working

yuzhouw313 commented on utils/michigan.py:
```diff
@@ -0,0 +1,530 @@
+
+    # columns to be added {full_name, entity_type, state, party, company,
+    # transaction_id,
+    #  "donor_id", ""}
```
In the clean.py, instead of creating an entity mapping dictionary property, I think we need to create a function in which we create that dictionary, I think we need to follow that structure in our state cleaner?

necabotheking replied:
When did we decide on that? My understanding was that the dictionary was just going to be initiated like this?

necabotheking commented on utils/michigan.py:
```diff
@@ -0,0 +1,530 @@
+    MICHIGAN_CONTRIBUTION_COLS_RENAME,
+    MICHIGAN_CONTRIBUTION_COLS_REORDER,
+)
+from utils.preprocess_mi_campaign_data import (
```
Are we not keeping these separate to use within the EDA too?

necabotheking commented on utils/michigan.py:
```diff
@@ -0,0 +1,530 @@
+        contribution_dataframe = self.merge_dataframes(temp_cont_list)
+        expenditure_dataframe = self.merge_dataframes(temp_exp_list)
+
+        return [contribution_dataframe, expenditure_dataframe]
```
There isn't a clean way to merge these files in preprocess, clean would be the only place I can succinctly merge the dataframes

necabotheking commented on utils/michigan.py:
```diff
@@ -0,0 +1,530 @@
+                "fundraiser",
+            ]
+        )
+        merged_expenditure_dataframe["full_name"] = (
```
If the contribution and expenditure datasets are going to be merged, then wouldn't a helper function that returns this data be unecessary?

averyschoen replied:
correct

necabotheking commented on utils/michigan.py:
```diff
@@ -0,0 +1,913 @@
+
+            # Map the generated UUIDs to a new column in the DataFrame
+            merged_campaign_dataframe[
+                "{}_uuid".format(col_name)
```
@trevorspreadbury @averyschoen Just to confirm, I originally planned to do this and mentioned this in a Mentor meeting but was told to create unique IDs for each entity in the data frame. Could you please confirm this.

necabotheking commented:
I addressed most of the feedback above with either questions, comments, or code fixes. The code works by calling clean_state() and outputs a 1GB Michigan ID MAP csv. I would love another review of the code to make it more efficient if you have time before Monday @trevorspreadbury @averyschoen

averyschoen commented on utils/michigan.py:
```diff
@@ -0,0 +1,1052 @@
+        """
+        merged_dataframe = standardized_dataframe_lst[0]
+
+        individuals_df = self.filter_dataframe(merged_dataframe, "first_name")
```
Why are you breaking out individuals vs candidates? A candidate should be an individual... collapsing this would help save quite a few lines and make this more legible

averyschoen commented on utils/michigan.py:
```diff
@@ -0,0 +1,1052 @@
+
+        return filtered_df
+
+    def filter_irrelevant_columns(
```
I don't think this is a necessary helper function

necabotheking replied:
The same piece of code is used in multiple places. Do you want me to get rid of it?

averyschoen commented on utils/michigan.py:
```diff
@@ -0,0 +1,1052 @@
+
+        return filtered_dataframe
+
+    def reorder_columns(
```
I also don't think this is a necessary helper function

necabotheking replied:
same comment as above: The same piece of code is used in multiple places. Do you want me to get rid of it?

averyschoen commented on utils/michigan.py:
```diff
@@ -0,0 +1,1052 @@
+        # convert committee IDs to integer, amount and aggregate cols to float
+        merged_contribution_dataframe = merged_contribution_dataframe.drop(
+            columns=[
+                "doc_seq_no",
```
you could also put these lists in constants to get rid of ~30 lines between contribution and expenditure functions

averyschoen commented on utils/michigan.py:
```diff
@@ -0,0 +1,1052 @@
+        michigan_id_map.to_csv(output_path, index=False)
+
+    def create_individuals_id_mapping(
+        self, individuals: pd.DataFrame, candidates: pd.DataFrame
```
very confused why this is split out

necabotheking replied:
Reference to the other comment above.

 In the dataframe individual donors and organizational donors are in the same columns (first_name, last_name) etc the only way to tell them apart is if the first_name is null. The candidates data is only in the expenditures column Which is why this has to be broken out into individuals and candidates then concatenated. Adiditionally, to create the id_mapping in the structure that yall wanted I had to create this function

averyschoen commented on utils/michigan.py:
```diff
@@ -0,0 +1,1052 @@
+
+        return id_mapping
+
+    def create_organizations_id_mapping(
```
also unclear why this is split out

necabotheking replied:
the id_mapping for each of the tables couldn't be done at once since the year columns were dropped before then per the schema. So I had to add the id mapping into create_tables(). Please let me know if there is a better way to do this

necabotheking replied:
It would essentially take the same code that's written above to do that. The whole process would just be moved into the transactions functions. Should i do that before Thursday then?

averyschoen commented on utils/michigan.py:
```diff
@@ -0,0 +1,1052 @@
+        """Creates the ID mapping dataframe for transactions
+
+        Inputs:
+            org_com: dataframe with organizations to committee transactions
```
WHAT????? the dfs should be org->org, org->ind, ind->ind, ind->org

averyschoen reviewed (changes requested):
I am very confused with what is going on in michigan.py. I hope my comments are helpful but I'm also around on tomorrow and Monday if you would like to discuss.

necabotheking commented on utils/michigan.py:
```diff
@@ -0,0 +1,1052 @@
+        "contribtype": "transaction_type",
+        "schedule_desc": "transaction_type",
+    }
+    # individuals_column_order = []
```
As a reply to your comment above, (and I also commented this above). I tried making them methods of the class and they broke the entire cleaner for some reason

necabotheking commented on utils/michigan.py:
```diff
@@ -0,0 +1,1052 @@
+        """
+        filepath_lst = []
+
+        for file in MI_EXP_FILEPATH.iterdir():
```
I plan to rename the new filepaths to be MI_EXP_FILEPATH since the old filepaths will not be needed
